enhancements/coreos-encrypted-by-default.md

```diff
@@ -0,0 +1,422 @@
+  - TBD
+approvers:
+  - TBD
+creation-date: 2019-09-20
+last-updated: 2019-09-20
+status: provisional
+---
+
+# CoreOS-Encrypted-by-Default
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs]
+- [ ] Transition Plan defined for Ignition handling of LUKS root volumes
+
+## Summary
+
+Openshift clusters will be encrypted at the operating system level by default. CoreOS boot images starting with 4.3 will be encrypted using a null-cipher. At boot time, the instance will be re-encrypted using AES-256-CBC and will not require user-intervention for boot and reboot.
+
+## Motivation
+
+The security of data-at-rest is of chief concern for end-users, organizations and enterprise Information Technology management. Every year, familiar companies have severe data breaches. Some industries require the encryption of data-at-rest, while customers of those industries demand it. Encryption of data is a best practice.
+
+### Goals
+
+This enhancement is to provide enterprise-grade encryption by default at the operating system level and to provide flexibility regarding the configuration.
+
+This means:
+* On first-boot, boot images based on CoreOS will be encrypted using standard AES-256 encryption.
+* OS-Level hooks will be included to support this feature
+* Automated boot/reboot handled by Clevis using "pins" for
+    * TPM2
+    * Tang
+    * Cloud KMS build around the CoreOS CryptAgent
+    * plaintext to support pass-through
+* Use of FIPS 140-2 compliant ciphers, hashes and checksum algorithms.
+* User intervention MUST not be required to initialize or (re)boot a cluster.
+* Users will have the ability to opt-out of encryption by default.
+
+### Non-Goals
+
+The following are considered out of scope:
+* Encryption of installations not based on a compatible boot image.
+* Changing Ignition.
+* Installer support.
+* Storing of keys in-Cluster.
+* Arbitrary LUKS configurations or disk-layouts.
+* Re-formatting disk devices or partitions.
+
+### Fedora (FCOS) and Red Hat CoreOS (RHCOS)
+
+It is worth stating that this proposal is not mutually exclusive for any plans that FCOS may have towards root disk layout. FCOS and RHCOS would like to reach alignment and feature parity.
+
+With the exclusion of the on-disk format, this work should lay the groundwork for an unattended encrypted boot that can be used by both FCOS and RHCOS.
+
+Ignition is the first boot agent for FCOS and RHCOS. FCOS discussed the topic of partition layouts, with the idea of having Ignition lay-down arbitrary root filesystem layouts. At this time, Openshift requires Ignition V2, while FCOS uses Ignition v3; these versions are incompatible. When Ignition supports the required depenendencies, a reviewion to this proposal will be forthcoming.
```

Comment:
Hmm.  I think though 90% of the code would be shared for implementing generic rootfs rewrites.  I don't think it needs to hard-depend on spec3 although it'd be nice.

We should at least *consider* making this depend on a spec3 transition, as it's currently the most significant delta between FCOS and RHCOS.

Reply:
The hard problem is teaching Ignition setting up the root disk. In the prep-work for this, the Ignition folks have concerns about teaching Ignition to change the root filesystem. Sure 90% of the code doing a rpm-ostree copy to memory and then re-splatting that to disk is reusable, we would need to update Spec v3 to support it.